pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -615,3 +615,35 @@ def is_module(self) -> bool:
         from pytorch_lightning import LightningDataModule, LightningModule  # prevent cyclic import
 
         return isinstance(self.instance, (LightningModule, LightningDataModule))
+
+
+@dataclass
+class _DataHookSource:
+    """Stores the info about the shared DataHooks within LightningModule and LightningDataModule.
+
+    The hook source can be
+
+    1. a method from the :class:`~pytorch_lightning.core.lightning.LightningModule`,
+    2. a method from the :class:`~pytorch_lightning.core.datamodule.LightningDataModule`,
+
+    Arguments:
+        model: A LightningModule
+        datamodule: A LightningDataModule
+    """
+
+    model: Optional["pl.LightningModule"]
+    datamodule: Optional["pl.LightningDataModule"]
+
+    def get_hook(self, hook_name):
+        if self.datamodule is None:
+            return getattr(self.model, hook_name)
+
+        if is_overridden(hook_name, self.datamodule):
+            if is_overridden(hook_name, self.model):
+                warning_cache.warn(
+                    f"You have overridden `{hook_name}` in both LightningModule and LightningDataModule."
+                    " We will use the implementation from LightningDataModule instance."
+                )
+            return getattr(self.datamodule, hook_name)
+
+        return getattr(self.model, hook_name)
```

Comment:
For datamodule is not None, but the specific hook function has not been overridden, current logic is to use hooks from model. Is this a valid behavior? How about throwing a warning that requested hooks come from model instead of datamodule? Because *_dataloader hook source may come from datamodule, but the transfer_* hook may come from lightningmodule.

Reply:
I kept that for BC since there's no warning in such a case as of now. Do you think we should change this behavior and mention it in the warning that if datamodule is passed then it will strictly use datamodule hooks only in v1.7?

or just a warning is fine?

Reply:
I'm fine with just warning. User can specify specific data source in `.fit()` , but it did not specify where the transfer hook source come from, so it will be better to communicate and clarifying the underlying logics.